### ISSUE TITLE
Using MSBuild custom toolchains

### ISSUE THREAD

**Issue by GCourtney27** [premake/premake-core#1590]
Are custom toolchains supported in premake? Say for adding targets to MSBuild. If so how would you go about adding them to premake?

**Comment by samsinsane** [comment]
We have support for modules that can extend the VS module. You can see how [Android](https://github.com/premake/premake-core/tree/master/modules/android) support is done, this is a module within the core library and works slightly differently to a module outside of the core library. For example, here's a module I wrote for targeting the [VSLinux](https://github.com/LORgames/premake-vslinux) extension. Is this the kind of thing you're looking for?

**Comment by GCourtney27** [comment]
I was thinking more along the lines of adding additional build targets with .target files. Like [this](http://www.reedbeta.com/blog/custom-toolchain-with-msbuild/) tutorial shows.

**Comment by samsinsane** [comment]
I've never had to use something like this before, so my knowledge about how Premake can handle this is going to be limited.

We have support for creating [custom rules](https://github.com/premake/premake-core/wiki/Custom-Rules) but that doesn't quite look like what you want since this will create the `.target` file as well as include it in the `.vcxproj`. We also have [`buildcustomizations`](https://github.com/premake/premake-core/wiki/buildcustomizations) (hasn't been documented yet), however, this requires the target file to exist relative to the `$(VCTargetsPath)`, this is what our internal code is doing: https://github.com/premake/premake-core/blob/1b7d93e6d3b7bb33a89c8a854a96691180c29385/modules/vstudio/vs2010_vcxproj.lua#L2027-L2031

Failing this, you can always workaround this limitation by doing:
```lua
premake.override(premake.vstudio.vc2010, "importExtensionTargets", function (oldfn, prj)
  oldfn(prj)
  premake.w('<Import Project="filename.targets" />')
end)
```
I think that's in the correct location in the `.vcxproj` structure based on the article you linked.

**Comment by GCourtney27** [comment]
So if I had a `.targets` file in "solutiondir/Some/Folder/build.targets" would I use `buildcustomizations` like `buildcustomizations {"build"}` after I do 
```
premake.override(premake.vstudio.vc2010, "importExtensionTargets", function (oldfn, prj)
  oldfn(prj)
  premake.w('<Import Project="$(SolutionDir)\\Some\\Folder\\%s.targets" />')
end)
```
in the premake file?

**Comment by samsinsane** [comment]
No, you would just use the `premake.override` and where you have `%s.targets` you should have `build.targets`. This override is just allowing you to write directly to all of the `.vcxproj` files that are generated, it's not limited to any specific project regardless of where you put it in your Premake files, and it's not using any of the existing APIs for input.
